pkg/replica/hash.go

```diff
@@ -0,0 +1,340 @@
+package replica
+
+import (
+	"bytes"
+	"context"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"hash"
+	"hash/crc64"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/gofrs/flock"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/sys/unix"
+
+	diskutil "github.com/longhorn/longhorn-engine/pkg/util/disk"
+)
+
+const (
+	defaultHashMethod = "crc64"
+
+	xattrSnapshotHashName     = "user.longhorn.metadata"
+	xattrSnapshotHashValueMax = 256
+
+	fileLockDirectory = "/host/var/lib/longhorn/.lock"
+	hashLockFileName  = "hash"
+)
+
+type SnapshotHashStatus struct {
+	StatusLock sync.RWMutex
+
+	State             ProgressState
+	Checksum          string
+	Error             string
+	SilentlyCorrupted bool
+}
+
+type SnapshotHashTask struct {
+	sync.Mutex
+
+	Ctx        context.Context
+	CancelFunc context.CancelFunc
+
+	SnapshotName string
+	Rehash       bool
+
+	file *os.File
+
+	SnapshotHashStatus
+}
+
+type SnapshotXattrHashInfo struct {
+	Method       string `json:"method"`
+	Checksum     string `json:"checksum"`
+	ModTime      string `json:"modTime"`
+	LastHashedAt string `json:"lastHashedAt"`
+}
+
+func NewSnapshotHashTask(ctx context.Context, cancel context.CancelFunc, snapshotName string, rehash bool) *SnapshotHashTask {
```

Comment:
NIT: it's better to rename this. `task` in this repo typically means the struct in `pkg/sync/sync.go`. This name is a little bit misleading...